Fix hdfs-related IT tests for java24

ldematte opened the pull request:
HDFS IT tests started failing with Java 24.

The problem is that some hadoop libraries (mainly `hdfs-common`) contain a call to a SecurityManager related method that has been removed in JDK 24.
A short term solution is to patch `hdfs-fixture` as we do for `hadoop-client-api` (same as we did in https://github.com/elastic/elasticsearch/pull/119779)

Fixes https://github.com/elastic/elasticsearch/issues/121967
Fixes https://github.com/elastic/elasticsearch/issues/122377
Fixes https://github.com/elastic/elasticsearch/issues/122378

ldematte committed:
Use hadoop patcher for hdfs-fixture too

ldematte committed:
Merge remote-tracking branch 'upstream/main' into fix-hdfs-tests-java24

ldematte commented on test/fixtures/hdfs-fixture/build.gradle:
```diff
@@ -104,6 +120,7 @@ def hdfs2Jar = tasks.register("hdfs2jar", ShadowJar) {
   }
   archiveClassifier.set("hdfs2")
   from sourceSets.main.output
+  from patchTask
```
Should we also exclude the non-patched classes here?
The generated JAR is OK, it contains the patched classes and I see no error, but I don't know if we need to be more explicit here.

ldematte commented on plugins/repository-hdfs/hadoop-common-patcher/src/main/java/org/elasticsearch/hdfs/patch/HdfsClassPatcher.java:
```diff
@@ -41,18 +41,17 @@ public static void main(String[] args) throws Exception {
             for (var patcher : patchers.entrySet()) {
                 JarEntry jarEntry = jarFile.getJarEntry(patcher.getKey());
-                if (jarEntry == null) {
-                    throw new IllegalArgumentException("path [" + patcher.getKey() + "] not found in [" + jarPath + "]");
```
This had to go, as some classes are not present in both jars; it's unfortunate, but I think acceptable. If it's not, should we pass the classes to patch via the command line, so it is configurable in each gradle `JavaExec` task?

rjernst replied:
we should pass the files to patch down, it should be part of the configuration of the task. we can keep how to actually patch the classes here, but the list of classes to patch be passed in.

However, why would the same class be present in two different jars? That's jarhell, which we should be rejecting already?

ldematte replied:
I should have said both _projects_; the hdsf-fixture projects needs a subset of the patches, as it pulls in a different (smaller) set of dependencies.

rjernst replied:
I'm still missing something. Why do dependencies matter? We should only be patching a given class once.

rjernst reviewed (commented):
I'll let @mark-vieira comment on the gradle, but I suspect we may want this to be a gradle plugin so we don't eg have to duplicate the patchClasses task and setup.

breskeby commented:
@ldematte This is more or less the perfect use case to use [Gradle Artefact Transforms](https://docs.gradle.org/current/userguide/artifact_transforms.html#sec:artifact-transforms-overview). The goal of those artefact transforms is to tweak third party dependencies before using them in your build. 

The benefits in general are IMO
- the transformations is done transparent to the usage as part of the dependency resolution
- we have built in caching support
- less clutter and custom build logic in our gradle files.
- is also that caching is build in supported and the project using these transforms is cleaner. 
- Also we probably can probably broaden its usage and reusability in the future as we 'patch' other dependencies as well (e.g. jwt)

I pushed a draft on how to port this PR to use artefact transforms to https://github.com/breskeby/elasticsearch/tree/fix-hdfs-tests-java24-artifact-transforms 

Happy to talk about this synchronously and help bring that over the line.

breskeby reviewed (changes requested):
I strongly prefer using gradle artefact transforms for these kind of dependency patching. Details at https://github.com/elastic/elasticsearch/pull/122044#issuecomment-2650912505

ldematte commented:
Thanks Rene, I'll update it as we discussed and make it parametric, so we can have back the check that the expected classes are all patched.

rjernst commented:
@breskeby What does the resulting artifact naming look like? One nice thing currently about the way we do this kind of patching is that we can make clear it's elastic-modified (not that we _do_ change the name in this case currently, but in other cases eg with log4j we have tweaked the name).

ldematte committed:
Merge remote-tracking branch 'upstream/main' into fix-hdfs-tests-java24

ldematte committed:
Try to use gradle transformer, not working

ldematte committed:
Removing patcher project

ldematte commented on plugins/repository-hdfs/hadoop-client-api/build.gradle:
```diff
@@ -2,53 +2,32 @@ import org.gradle.api.file.ArchiveOperations
-sourceSets {
-  patcher
-}
+def patched = Attribute.of('patched', Boolean)
```
I think this whole project is not needed anymore and we can transform directly in the repository-hdfs project (parent)

Commit:
[CI] Auto commit changes from spotless

ldematte committed:
Patcher fix

ldematte committed:
Merge branch 'fix-hdfs-tests-java24' of github.com:ldematte/elasticsearch into fix-hdfs-tests-java24

ldematte committed:
Merge remote-tracking branch 'upstream/main' into fix-hdfs-tests-java24

ldematte committed:
Fix repository-hdfs gradle to perform transformation directly

ldematte committed:
Add test classpaths

ldematte committed:
Merge remote-tracking branch 'upstream/main' into fix-hdfs-tests-java24

ldematte committed:
Unmuting

elasticsearchmachine commented:
Pinging @elastic/es-core-infra (Team:Core/Infra)

elasticsearchmachine commented:
Pinging @elastic/es-delivery (Team:Delivery)

ldematte commented:
I moved the patcher implementation to a transformer as @breskeby suggested, and added the ability to specify (via a parameter) which set of classes to patch for which jar/artifact.

breskeby commented on build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/dependencies/patches/hdfs/HdfsClassPatcher.java:
```diff
@@ -0,0 +1,126 @@
+                // Add the entry to the new JAR file
+                jos.putNextEntry(new JarEntry(entryName));
+
+                Function<ClassWriter, ClassVisitor> classPatcher = jarPatchers.get(entryName);
```
I wonder if we should add a check here to verify all expected patches are applied and fail early if not to avoid having undetected changes sneaked into this again by e.g. updating hadoop versions.

ldematte replied:
I tried to do that and it's a great idea; however there is one big blocker for this: the hadoop-common artifact has one additional "classifier" named `tests`, which contains only tests of course but is brought in anyway.
I tried to exclude it but apparently it's not possible to do that in gradle (there is limited support for classifiers).
The thing is that hadoop-common is applied to both hadoop-common jars (regular and "tests"), and it will fail on the second, because none of the expected patches will be applied there.

ldematte replied:
TL;DR: if you have a good way to either exclude `hadoop-common:tests` or to recognize it and avoid trying to patch it I'm all ears!

breskeby replied:
We could tweak the regex to match the file name being stricter in a sense of expected name + version regex instead of just a simple string.contains based check to fix to not match -tests artifacts

ldematte replied:
Let me try and see the regex

breskeby replied:
I had to deal with that test classifier stuff with hadoop before and suffered myself. okay. lets just live with that for now. we can revisit later if we feel the need. I will use parts of that stuff later on to fix other places in our codebase.

breskeby commented on plugins/repository-hdfs/build.gradle:
```diff
@@ -69,6 +84,20 @@ dependencies {
   hdfsFixture2 project(path: ':test:fixtures:hdfs-fixture', configuration: 'shadowedHdfs2')
   hdfsFixture3 project(path: ':test:fixtures:hdfs-fixture', configuration: 'shadow')
+
+  attributesSchema {
```
nit: can we put that attributesSchema, attributeTypes and transform definitions on top of the dependency{} block?

breskeby commented on test/fixtures/hdfs-fixture/build.gradle:
```diff
@@ -13,8 +13,29 @@ apply plugin: 'com.gradleup.shadow'
 import com.github.jengelman.gradle.plugins.shadow.tasks.ShadowJar
 
+def patched = Attribute.of('patched', Boolean)
+dependencies {
```
nit. the usual order in a build script is configuration definitions before dependencies. kind of best practice we usually follow. so ideally we would put that dependencies{} block after the configurations{} block

breskeby reviewed (approved):
This is great and getting rid of hadoop-client-api subproject is a very nice sideeffect. I added a few nitpicks which would be great to be addressed before merging. 

Otherwise LGTM.

ldematte committed:
Reorder gradle file

ldematte committed:
Regex?

ldematte committed:
Merge branch 'main' into fix-hdfs-tests-java24

ldematte committed:
Add hdfs3 fixture too

ldematte committed:
Use regex to match artifact; add check that we patch all classes in the matched artifacts

ldematte committed:
Merge branch 'fix-hdfs-tests-java24' of github.com:ldematte/elasticsearch into fix-hdfs-tests-java24

breskeby commented on build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/dependencies/patches/hdfs/HdfsClassPatcher.java:
```diff
@@ -0,0 +1,147 @@
+    static final List<JarPatchers> allPatchers = List.of(
+        new JarPatchers(
+            "hadoop-common",
+            Pattern.compile("hadoop-common-(?!.*tests)"),
```
lgtm
